### PR TITLE
Fix some uTP issues

### PR DESCRIPTION
**Pull request by ogenev**
### What was wrong?
Fixes https://github.com/ethereum/trin/issues/417

Some uTP issues were observed when testing uTP protocol with 10k byte transfer and network delay.

### How was it fixed?
- Build selective ack for an empty incoming buffer, this case was not handled.
- Insert packets into the incoming buffer before packet handling.
- Remove all acknowledged packets from send window upon selective ack receive.
- Emit close event after FIN packet is fully handled.

### To-Do

[//]: # (Stay ahead of things, add list items here!)
- [x] Add entry to the [release notes](https://github.com/ethereum/trin/blob/master/newsfragments/README.md) (may forgo for trivial changes)
- [x] Clean up commit history

### CODE REVIEW

File: newsfragments/417.fixed.md [hunk 1]
```diff
@@ -0,0 +1,2 @@
+- Fix selective ask bug in uTP protocol.
```
Comment (jacobkaufmann) [inline newsfragments/417.fixed.md:1]:
```suggestion
- Fix selective ack bug in uTP protocol.
```

File: utp-testing/docker/circleci/Dockerfile [hunk 2]
```diff
@@ -1,4 +1,4 @@
-FROM debian:bullseye-slim
+FROM ubuntu:22.04
```
Comment (jacobkaufmann) [inline utp-testing/docker/circleci/Dockerfile:1]: Why the change here?

**Review by jacobkaufmann — COMMENTED** [review]
Nice! I would like to see some unit tests alongside the changes that cover these bugs.

File: utp-testing/docker/circleci/Dockerfile (hunk 2 above)
Comment (ogenev) [reply utp-testing/docker/circleci/Dockerfile:1]: There are some issues with the Debian slim image when I try to run the test app container locally. Switching to the latest ubuntu LTS works like a charm.

File: trin-core/src/utp/stream.rs [hunk 3]
```diff
@@ -320,9 +320,18 @@ impl UtpListener {
...
+                            let mut result = Vec::new();
                             let mut buf = [0; BUF_SIZE];
-                            if let Err(msg) = conn.recv(&mut buf).await {
-                                error!("Unable to receive uTP FIN packet: {msg}")
+
+                            match conn.recv(&mut buf).await {
+                                Ok(bytes_read) => {
+                                    if let Some(bytes) = bytes_read {
+                                        result.extend_from_slice(&buf[..bytes]);
+                                        conn.recv_data_stream.append(&mut result);
```
Comment (jacobkaufmann) [inline trin-core/src/utp/stream.rs:330]: Why do we need `result` here? We can just do the following:

```rust
conn.recv_data_stream.extend_from_slice(&buf[..bytes]);
```

File: trin-core/src/utp/stream.rs [hunk 4]
```diff
@@ -789,6 +798,14 @@ impl UtpStream {
 
     /// Builds the selective acknowledgement extension data for usage in packets.
     fn build_selective_ack(&self) -> Vec<u8> {
+        // Build selective ack for empty incoming buffer
+        if self.incoming_buffer.is_empty() {
+            let mut sack: Vec<u8> = vec![0u8; 4];
+            sack[0] |= 1 << 0;
```
Comment (jacobkaufmann) [inline trin-core/src/utp/stream.rs:804]:
```suggestion
            let mut sack: Vec<u8> = vec![0u8; 4];
            sack[0] = 1;
```

File: trin-core/src/utp/stream.rs [hunk 5]
```diff
@@ -1097,6 +1113,25 @@ impl UtpStream {
...
+                        .take_while(|&seq_nr| seq_nr <= last_seq_nr);
+
+                    for seq_nr in ack_packets {
+                        if let Some(position) = self
+                            .send_window
+                            .iter()
+                            .position(|packet| packet.seq_nr() == seq_nr)
+                        {
+                            let packet = self.send_window.remove(position);
+                            self.cur_window = self.cur_window.saturating_sub(packet.len() as u32);
```
Comment (jacobkaufmann) [inline trin-core/src/utp/stream.rs:1131]: Should we be panicking here instead of `saturating_sub`? If we overflow, does that reflect some incorrect accounting on our part, or is this defense from some incorrect behavior from the peer?

Comment (ogenev) [reply trin-core/src/utp/stream.rs:1131]: Thinking about it now, panicking makes more sense because it will indicate a bug in our code.

**Review by njgheorghita — APPROVED** [review]
:ship:
